### DISCUSSION TITLE
Status of the NLP implementation / What will come in the near future

### DISCUSSION THREAD

**Discussion by patrickgold** [florisboard/florisboard#2197] [Announcements]
**Please see the up-to-date state of the project here:** https://github.com/florisboard/florisboard/discussions/2314

-----

**TL;DR at the bottom**

-----

Hey everyone,

quite a long time has passed since the last update to FlorisBoard was made and I must admit it kinda seemed this project is abandoned from the outside due to the unfortunate lack of communication from my side which I must apoligize for. For the future I plan to better communicate the status of the project's development.

While it is true that due to private reasons I had to take a bit of a pause of developing FlorisBoard and some time passed with no progress at all, implementing a completely new statistical NLP (Natural Language Processing) provider, or in laymans terms the long-awaited word prediction and spell-checking implementation, is also a huge task which takes a lot of time and trial-error and development time. This announcement thus serves as a transparent report on my NLP implementation progress and if you are an advanced user you can already play around with it starting from today (more on that later though).

## (Technical) Details on the new NLP implementation

If you are a long-time FlorisBoard user you might remember on my first attempt on this NLP provider - And well it didn't go well. It was implemented in Kotlin, which meant it suffered from the performance penalties and memory restrictions of the JVM (Java Virtual Machine), which all Android apps by default run in. The algorithm was slow, did not provide good results, and the worst - it frequently produced those infamous OOM (OutOfMemory) crashes. A lot of cleanup had to be done in the issues and also a lot of the time the same error caused other components to go down as well.

While the previous attempt was a complete failure we can at least learn from the mistakes made and build a completely new, fast, reliable, and most-important stable NLP provider this time. To ensure that the NLP provider runs well I decided to implement it in C++. Also to be able to debug it and to preprocess dictionaries easily I decided to split the NLP provider codebase off into a separately repository, which you can view here:

https://github.com/florisboard/nlp

This repository contains the core NLP implementation to be used in the keyboard as well as tools for debugging and preprocessing on desktops. The repository will be used from now on to host any native code regarding NLP for FlorisBoard and will be included via a git submodule in the main project.

For the keyboard logic (which is written in Kotlin) to be able to access the new NLP implementation we need to bridge the two runtimes, so they can communicate flawlessly. This will be done in the main FlorisBoard repository.

## What's the status of the NLP implementation?

Above mentioned NLP core is in a state where it is ready to provide word predictions and spelling corrections, even if it still needs a lot of refinement based on user feedback.

The focus is now moving on building the aforementioned "bridge" as well as getting the whole code base to properly compile for all Android targets again, so I can ship the NLP core together with the APK. This part will be done in PR #2196 and you will be able to track the progress live again - unlike with my NLP core implementation which was private for the last 6 months or so.

### To take away the most common question I think I will get: When can we try out the new NLP implementation?

If you are a casual non-technical user: Not yet sadly. As said I still need to work on integrating the NLP core into the main project, but as this very important milestone was now reached I am confident that it will probably take a few more weeks until I can provide a new alpha release!

If you are a technical user or someone who likes to tinker with software a lot: You can absolutely try out the implementation starting from today, however please note that you need a Linux Desktop and probably a bit patience for the first build from zero. Head to the [NLP repository](https://github.com/florisboard/nlp) and check out the README!

### Some additional thoughts

Although I put a lot of work into this NLP provider bear in mind that it is an experimental **statistical** NLP model and some word predictions will look bizarre. Also atm ngrams are not pretrained and only learn live from user input, so it takes some time until the model can somewhat reliably predict the next word. If you are used to word prediction from modern AI models or even from ChatGPT you will be disappointed.

## A word on open issues

I am aware that during the last few months open issues have stockpiled and I really need to play catch-up here. At first I will prioritze open issues that are releated to the NLP implementation and on issues that severly impact the usability of the keyboard in general. I hope you can understand!

## End words

If you've come this far thanks for reading through this announcement! The nearer time frame will bring exiting improvements for this project and after the NLP provider has been implemented successfully I can focus again on ther features which are due for implementation.

~ Patrick

-----

**TL;DR**

The NLP implementation (word prediction + spell check) has reached an important milestone: The NLP core implemenation (which is hosted in a [separate repository](https://github.com/florisboard/nlp)) has reached a usable state and has now been pusblished. In the following weeks work will be focusing on bridging the NLP interface of the keyboard logic with the NLP core and then to be able to ship a new alpha release with working word prediction/spell checking. Progress on this can be tracked in PR #2196

**Comment by msfjarvis** [comment]
Glad to have an update! With the continual ruin being brought to SwiftKey and Gboard, I've been recommending FlorisBoard increasingly often and both the lack of predictive text as well as general radio silence often came up as potential discouragements :)

**Comment by Kreuger** [comment]
Happy to continue testing the dev builds once we get to that place. Im still currently using the "beta" build but have the main and debug installed as well.

**Comment by HatakeKakashri** [comment]
@patrickgold @BHydden The nlp PR #2196 seems to be merged into the master. Can we expect alpha5 in the coming weeks.

**Reply by BHydden** [reply]
No need to tag me, I have no connection to the project beyond being an enthusiastic fan and curious onlooker. I'm not connected with the code in any way.

But to answer your question, probably shouldn't expect alpha5 until part two is also finished 

https://github.com/florisboard/florisboard/pull/2209

Hopefully Soon™ but no time frame has or will be given. When patrick feels it's ready for alpha testing he'll release it, right now there are still known bugs he wants to sort out.

**Reply by patrickgold** [reply]
Exactly as BHydden said, part two is crucial for NLP to work in FlorisBoard, and there are some really annoying bugs yet to be solved (e.g. build issues, performance issues, dictionary size, learning algorithm). Not everything needs to be flawlessly solved before I can make an alpha05 release, but at least the performance and the internal algorithms should kinda work IMO. And as annoying as it is I won't give any ETA because I just can't predict how much longer it takes me until I can finally release alpha05.

**Comment by pixincreate** [comment]
Just a thought, since we are already in the age of AI, can we use LLaMa 2.0 to make predictions possible? A small fine tuned model would be more than enough to make word predictions and spell check, right?

**Reply by Atemu** [reply]
Even the smallest llama models are *humongous* in mobile terms. That's not feasible right now.

**Reply by patrickgold** [reply]
Sadly I have to agree with Atemu here, atm those new fancy AI models are huge even in the smallest config, so it's not really feasible to implement right now. However generally a machine learning model could definitely help in achieving the goal of word prediction in the long term.
